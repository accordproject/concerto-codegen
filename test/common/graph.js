/* eslint-disable no-unreachable */
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const { DirectedGraph, ConcertoGraphVisitor } = require('../../lib/common/common.js');
const { ModelManager } = require('@accordproject/concerto-core');
const fs = require('fs');
const { expect } = require('expect');

const chai = require('chai');
const { InMemoryWriter } = require('@accordproject/concerto-util');
chai.should();
chai.use(require('chai-as-promised'));
chai.use(require('chai-things'));

describe('graph', function () {
    let modelManager = null;

    before(function() {
        process.env.ENABLE_MAP_TYPE = 'true'; // TODO Remove on release of MapType
    });

    beforeEach(function() {
        modelManager = new ModelManager();
        const hrBase = fs.readFileSync('./test/codegen/fromcto/data/model/hr_base.cto', 'utf-8');
        modelManager.addCTOModel(hrBase, 'hr_base.cto');
        const hr = fs.readFileSync('./test/codegen/fromcto/data/model/hr.cto', 'utf-8');
        modelManager.addCTOModel(hr, 'hr.cto');
    });


    describe('#visitor', function () {
        it('should visit a model manager', function () {
            const visitor = new ConcertoGraphVisitor();
            visitor.should.not.be.null;
            const writer = new InMemoryWriter();

            const graph = new DirectedGraph();
            modelManager.accept(visitor, { graph });

            writer.openFile('graph.mmd');
            graph.print(writer);
            writer.closeFile();
            expect(writer.data.get('graph.mmd')).toEqual(`flowchart LR
   \`org.acme.hr.base@1.0.0.State\`
   \`org.acme.hr.base@1.0.0.State\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.TShirtSizeType\`
   \`org.acme.hr.base@1.0.0.TShirtSizeType\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\`
   \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\` --> \`org.acme.hr.base@1.0.0.TShirtSizeType\`
   \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr.base@1.0.0.Address\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.Address\` --> \`org.acme.hr.base@1.0.0.State\`
   \`org.acme.hr.base@1.0.0.Time\`
   \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.CompanyProperties\`
   \`org.acme.hr@1.0.0.CompanyProperties\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeLoginTimes\`
   \`org.acme.hr@1.0.0.EmployeeLoginTimes\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeLoginTimes\` --> \`org.acme.hr.base@1.0.0.Time\`
   \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\`
   \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.NextOfKin\`
   \`org.acme.hr@1.0.0.NextOfKin\` --> \`org.acme.hr@1.0.0.KinName\`
   \`org.acme.hr@1.0.0.NextOfKin\` --> \`org.acme.hr@1.0.0.KinTelephone\`
   \`org.acme.hr@1.0.0.EmployeeProfiles\`
   \`org.acme.hr@1.0.0.EmployeeProfiles\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeProfiles\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.EmployeeDirectory\`
   \`org.acme.hr@1.0.0.EmployeeDirectory\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.EmployeeDirectory\` --> \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.Company\`
   \`org.acme.hr@1.0.0.Company\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.CompanyProperties\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.EmployeeDirectory\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.EmployeeProfiles\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\`
   \`org.acme.hr@1.0.0.Department\`
   \`org.acme.hr@1.0.0.Department\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Equipment\`
   \`org.acme.hr@1.0.0.Equipment\` --> \`concerto@1.0.0.Asset\`
   \`org.acme.hr@1.0.0.LaptopMake\`
   \`org.acme.hr@1.0.0.LaptopMake\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Laptop\`
   \`org.acme.hr@1.0.0.Laptop\` --> \`org.acme.hr@1.0.0.Equipment\`
   \`org.acme.hr@1.0.0.Laptop\` --> \`org.acme.hr@1.0.0.LaptopMake\`
   \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.Person\` --> \`concerto@1.0.0.Participant\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr@1.0.0.NextOfKin\`
   \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr@1.0.0.Department\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr@1.0.0.Equipment\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr@1.0.0.Manager\`
   \`org.acme.hr@1.0.0.Contractor\`
   \`org.acme.hr@1.0.0.Contractor\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.Contractor\` --> \`org.acme.hr@1.0.0.Company\`
   \`org.acme.hr@1.0.0.Contractor\` --> \`org.acme.hr@1.0.0.Manager\`
   \`org.acme.hr@1.0.0.Manager\`
   \`org.acme.hr@1.0.0.Manager\` --> \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.Manager\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.CompanyEvent\`
   \`org.acme.hr@1.0.0.CompanyEvent\` --> \`concerto@1.0.0.Event\`
   \`org.acme.hr@1.0.0.Onboarded\`
   \`org.acme.hr@1.0.0.Onboarded\` --> \`org.acme.hr@1.0.0.CompanyEvent\`
   \`org.acme.hr@1.0.0.Onboarded\` --> \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`concerto@1.0.0.Transaction\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.KinName\`
   \`org.acme.hr@1.0.0.KinTelephone\`
`);
        });


        it('should visit find a connected subgraph', function () {
            const visitor = new ConcertoGraphVisitor();
            visitor.should.not.be.null;
            const writer = new InMemoryWriter();

            const graph = new DirectedGraph();
            modelManager.accept(visitor, { graph });

            const connectedGraph = graph.findConnectedGraph('org.acme.hr@1.0.0.ChangeOfAddress');
            expect(connectedGraph.hasEdge('org.acme.hr@1.0.0.ChangeOfAddress', 'org.acme.hr@1.0.0.Person'));

            const filteredModelManager = modelManager
                .filter(declaration => connectedGraph.hasVertex(declaration.getFullyQualifiedName()));

            expect(filteredModelManager.getModelFiles()).toHaveLength(2);
            expect(filteredModelManager.getModelFiles()[0].getAllDeclarations()).toHaveLength(3);

            writer.openFile('graph.mmd');
            connectedGraph.print(writer);
            writer.closeFile();
            expect(writer.data.get('graph.mmd')).toEqual(`flowchart LR
   \`org.acme.hr.base@1.0.0.State\`
   \`org.acme.hr.base@1.0.0.State\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr.base@1.0.0.Address\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.Address\` --> \`org.acme.hr.base@1.0.0.State\`
   \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.NextOfKin\`
   \`org.acme.hr@1.0.0.NextOfKin\` --> \`org.acme.hr@1.0.0.KinName\`
   \`org.acme.hr@1.0.0.NextOfKin\` --> \`org.acme.hr@1.0.0.KinTelephone\`
   \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.Person\` --> \`concerto@1.0.0.Participant\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr@1.0.0.NextOfKin\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`concerto@1.0.0.Transaction\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.KinName\`
   \`org.acme.hr@1.0.0.KinTelephone\`
`);
        });
    });
});
