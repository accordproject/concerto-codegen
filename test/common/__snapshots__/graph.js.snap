// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph #visitor should visit a model manager 1`] = `
"flowchart LR
   \`org.acme.hr.base@1.0.0.Category\`
   \`org.acme.hr.base@1.0.0.Category\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr.base@1.0.0.GeneralCategory\` --> \`org.acme.hr.base@1.0.0.Category\`
   \`org.acme.hr.base@1.0.0.State\`
   \`org.acme.hr.base@1.0.0.State\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.TShirtSizeType\`
   \`org.acme.hr.base@1.0.0.TShirtSizeType\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\`
   \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\` --> \`org.acme.hr.base@1.0.0.TShirtSizeType\`
   \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr.base@1.0.0.Address\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.Address\` --> \`org.acme.hr.base@1.0.0.State\`
   \`org.acme.hr.base@1.0.0.Level\`
   \`org.acme.hr.base@1.0.0.Level\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.Time\`
   \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.pii\`
   \`org.acme.hr@1.0.0.pii\` --> \`concerto.decorator@1.0.0.Decorator\`
   \`org.acme.hr@1.0.0.pii\` --> \`org.acme.hr@1.0.0.Category\`
   \`concerto.decorator@1.0.0.Decorator\`
   \`org.acme.hr@1.0.0.Category\`
   \`org.acme.hr@1.0.0.Category\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Info\`
   \`org.acme.hr@1.0.0.Info\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Info\` --> \`org.acme.hr@1.0.0.pii\`
   \`org.acme.hr@1.0.0.Info\` --> \`org.acme.hr@1.0.0.Category\`
   \`org.acme.hr@1.0.0.CompanyProperties\`
   \`org.acme.hr@1.0.0.CompanyProperties\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeLoginTimes\`
   \`org.acme.hr@1.0.0.EmployeeLoginTimes\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeLoginTimes\` --> \`org.acme.hr.base@1.0.0.Time\`
   \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\`
   \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.NextOfKin\`
   \`org.acme.hr@1.0.0.NextOfKin\` --> \`org.acme.hr@1.0.0.KinName\`
   \`org.acme.hr@1.0.0.NextOfKin\` --> \`org.acme.hr@1.0.0.KinTelephone\`
   \`org.acme.hr@1.0.0.EmployeeProfiles\`
   \`org.acme.hr@1.0.0.EmployeeProfiles\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.EmployeeProfiles\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeProfiles\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.EmployeeDirectory\`
   \`org.acme.hr@1.0.0.EmployeeDirectory\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.EmployeeDirectory\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.EmployeeDirectory\` --> \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.Company\`
   \`org.acme.hr@1.0.0.Company\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.CompanyProperties\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.EmployeeDirectory\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.EmployeeProfiles\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\`
   \`org.acme.hr@1.0.0.Department\`
   \`org.acme.hr@1.0.0.Department\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Department\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.Equipment\`
   \`org.acme.hr@1.0.0.Equipment\` --> \`concerto@1.0.0.Asset\`
   \`concerto@1.0.0.Asset\`
   \`org.acme.hr@1.0.0.LaptopMake\`
   \`org.acme.hr@1.0.0.LaptopMake\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Laptop\`
   \`org.acme.hr@1.0.0.Laptop\` --> \`org.acme.hr@1.0.0.Equipment\`
   \`org.acme.hr@1.0.0.Laptop\` --> \`org.acme.hr@1.0.0.LaptopMake\`
   \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.Person\` --> \`concerto@1.0.0.Participant\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr@1.0.0.NextOfKin\`
   \`concerto@1.0.0.Participant\`
   \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr@1.0.0.Department\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr@1.0.0.Equipment\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.Contractor\`
   \`org.acme.hr@1.0.0.Contractor\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.Contractor\` --> \`org.acme.hr@1.0.0.Company\`
   \`org.acme.hr@1.0.0.Contractor\` --> \`org.acme.hr.base@1.0.0.Level\`
   \`org.acme.hr@1.0.0.Contractor\` --> \`org.acme.hr@1.0.0.Manager\`
   \`org.acme.hr@1.0.0.Manager\`
   \`org.acme.hr@1.0.0.Manager\` <--> \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.Manager\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.CompanyEvent\`
   \`org.acme.hr@1.0.0.CompanyEvent\` --> \`concerto@1.0.0.Event\`
   \`concerto@1.0.0.Event\`
   \`org.acme.hr@1.0.0.Onboarded\`
   \`org.acme.hr@1.0.0.Onboarded\` --> \`org.acme.hr@1.0.0.CompanyEvent\`
   \`org.acme.hr@1.0.0.Onboarded\` --> \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`concerto@1.0.0.Transaction\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`concerto@1.0.0.Transaction\`
   \`org.acme.hr@1.0.0.KinName\`
   \`org.acme.hr@1.0.0.KinName\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.KinTelephone\`
   \`org.acme.hr@1.0.0\`
   \`org.acme.hr@1.0.0\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
"
`;

exports[`graph #visitor should visit a model manager and create a dependency graph 1`] = `
"flowchart LR
   \`org.acme.hr.base@1.0.0.Category\`
   \`org.acme.hr.base@1.0.0.Category\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr.base@1.0.0.GeneralCategory\` --> \`org.acme.hr.base@1.0.0.Category\`
   \`org.acme.hr.base@1.0.0.State\`
   \`org.acme.hr.base@1.0.0.State\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.TShirtSizeType\`
   \`org.acme.hr.base@1.0.0.TShirtSizeType\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\`
   \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\` --> \`org.acme.hr.base@1.0.0.TShirtSizeType\`
   \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr.base@1.0.0.Address\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.Address\` --> \`org.acme.hr.base@1.0.0.State\`
   \`org.acme.hr.base@1.0.0.Level\`
   \`org.acme.hr.base@1.0.0.Level\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.Time\`
   \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.pii\`
   \`org.acme.hr@1.0.0.pii\` --> \`concerto.decorator@1.0.0.Decorator\`
   \`org.acme.hr@1.0.0.pii\` --> \`org.acme.hr@1.0.0.Category\`
   \`org.acme.hr@1.0.0.Category\`
   \`org.acme.hr@1.0.0.Category\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Info\`
   \`org.acme.hr@1.0.0.Info\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Info\` --> \`org.acme.hr@1.0.0.pii\`
   \`org.acme.hr@1.0.0.Info\` --> \`org.acme.hr@1.0.0.Category\`
   \`org.acme.hr@1.0.0.CompanyProperties\`
   \`org.acme.hr@1.0.0.CompanyProperties\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeLoginTimes\`
   \`org.acme.hr@1.0.0.EmployeeLoginTimes\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeLoginTimes\` --> \`org.acme.hr.base@1.0.0.Time\`
   \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\`
   \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.NextOfKin\`
   \`org.acme.hr@1.0.0.NextOfKin\` --> \`org.acme.hr@1.0.0.KinName\`
   \`org.acme.hr@1.0.0.NextOfKin\` --> \`org.acme.hr@1.0.0.KinTelephone\`
   \`org.acme.hr@1.0.0.EmployeeProfiles\`
   \`org.acme.hr@1.0.0.EmployeeProfiles\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.EmployeeProfiles\` --> \`String\`
   \`org.acme.hr@1.0.0.EmployeeProfiles\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.EmployeeDirectory\`
   \`org.acme.hr@1.0.0.EmployeeDirectory\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.EmployeeDirectory\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.EmployeeDirectory\` --> \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.Company\`
   \`org.acme.hr@1.0.0.Company\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.CompanyProperties\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.EmployeeDirectory\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr.base@1.0.0.EmployeeTShirtSizes\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.EmployeeProfiles\`
   \`org.acme.hr@1.0.0.Company\` --> \`org.acme.hr@1.0.0.EmployeeSocialSecurityNumbers\`
   \`org.acme.hr@1.0.0.Department\`
   \`org.acme.hr@1.0.0.Department\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Department\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.Equipment\`
   \`org.acme.hr@1.0.0.Equipment\` --> \`concerto@1.0.0.Asset\`
   \`org.acme.hr@1.0.0.LaptopMake\`
   \`org.acme.hr@1.0.0.LaptopMake\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr@1.0.0.Laptop\`
   \`org.acme.hr@1.0.0.Laptop\` --> \`org.acme.hr@1.0.0.Equipment\`
   \`org.acme.hr@1.0.0.Laptop\` --> \`org.acme.hr@1.0.0.LaptopMake\`
   \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.Person\` --> \`concerto@1.0.0.Participant\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr@1.0.0.NextOfKin\`
   \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr@1.0.0.Department\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr@1.0.0.Equipment\`
   \`org.acme.hr@1.0.0.Employee\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.Contractor\`
   \`org.acme.hr@1.0.0.Contractor\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.Contractor\` --> \`org.acme.hr@1.0.0.Company\`
   \`org.acme.hr@1.0.0.Contractor\` --> \`org.acme.hr.base@1.0.0.Level\`
   \`org.acme.hr@1.0.0.Contractor\` --> \`org.acme.hr@1.0.0.Manager\`
   \`org.acme.hr@1.0.0.Manager\`
   \`org.acme.hr@1.0.0.Manager\` <--> \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.Manager\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.CompanyEvent\`
   \`org.acme.hr@1.0.0.CompanyEvent\` --> \`concerto@1.0.0.Event\`
   \`org.acme.hr@1.0.0.Onboarded\`
   \`org.acme.hr@1.0.0.Onboarded\` --> \`org.acme.hr@1.0.0.CompanyEvent\`
   \`org.acme.hr@1.0.0.Onboarded\` --> \`org.acme.hr@1.0.0.Employee\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`concerto@1.0.0.Transaction\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.KinName\`
   \`org.acme.hr@1.0.0.KinName\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.KinTelephone\`
   \`org.acme.hr@1.0.0\`
   \`org.acme.hr@1.0.0\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
"
`;

exports[`graph #visitor should visit find a connected subgraph 1`] = `
"flowchart LR
   \`org.acme.hr.base@1.0.0.Category\`
   \`org.acme.hr.base@1.0.0.Category\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr.base@1.0.0.GeneralCategory\` --> \`org.acme.hr.base@1.0.0.Category\`
   \`org.acme.hr.base@1.0.0.State\`
   \`org.acme.hr.base@1.0.0.State\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr.base@1.0.0.Address\` --> \`concerto@1.0.0.Concept\`
   \`org.acme.hr.base@1.0.0.Address\` --> \`org.acme.hr.base@1.0.0.State\`
   \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.NextOfKin\`
   \`org.acme.hr@1.0.0.NextOfKin\` --> \`org.acme.hr@1.0.0.KinName\`
   \`org.acme.hr@1.0.0.NextOfKin\` --> \`org.acme.hr@1.0.0.KinTelephone\`
   \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.Person\` --> \`concerto@1.0.0.Participant\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr.base@1.0.0.SSN\`
   \`org.acme.hr@1.0.0.Person\` --> \`org.acme.hr@1.0.0.NextOfKin\`
   \`concerto@1.0.0.Participant\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`concerto@1.0.0.Transaction\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`org.acme.hr@1.0.0.Person\`
   \`org.acme.hr@1.0.0.ChangeOfAddress\` --> \`org.acme.hr.base@1.0.0.Address\`
   \`concerto@1.0.0.Transaction\`
   \`org.acme.hr@1.0.0.KinName\`
   \`org.acme.hr@1.0.0.KinName\` --> \`org.acme.hr.base@1.0.0.GeneralCategory\`
   \`org.acme.hr@1.0.0.KinTelephone\`
"
`;
